### PR TITLE
Porta cipher

### PR DESCRIPTION
**Pull request by mrvnmchm**
Added porta cipher with doctests for #1492

### CODE REVIEW

File: ciphers/porta_cipher.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+def generateTable(key):
+    """
+    >>> generateTable('marvin')
+    [[['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M'], ['U', 'V', 'W', 'X', 'Y', 'Z', 'N', 'O', 'P', 'Q', 'R', 'S', 'T']], [['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M'], ['N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z']], [['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M'], ['S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'N', 'O', 'P', 'Q', 'R']], [['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M'], ['Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'N', 'O', 'P']], [['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M'], ['W', 'X', 'Y', 'Z', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V']], [['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M'], ['U', 'V', 'W', 'X', 'Y', 'Z', 'N', 'O', 'P', 'Q', 'R', 'S', 'T']]]
+    """
+    table = []
+    for char in key.upper():
+        table.append(alphabet[char])
+    return table
```
Comment (cclauss) [inline ciphers/porta_cipher.py]: Please convert these four lines into a one line list comprehension.
`return [alphabet[char] for char in key.upper()]`

File: ciphers/porta_cipher.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+          ["R", "S", "T", "U", "V", "W", "X", "Y", "Z", "N", "O", "P", "Q"]], "T": [["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M"], ["R", "S", "T", "U", "V", "W", "X", "Y", "Z", "N", "O", "P", "Q"]],
+    "U": [["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M"],
+          ["Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z", "N", "O", "P"]], "V": [["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M"], ["Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z", "N", "O", "P"]],
+    "W": [["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M"],
+          ["P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z", "N", "O"]], "X": [["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M"], ["P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z", "N", "O"]],
+    "Y": [["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M"],
+          ["O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z", "N"]], "Z": [["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M"], ["O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z", "N"]]}
+
+
+def generateTable(key):
```
Comment (cclauss) [inline ciphers/porta_cipher.py]: `def generate_table(key):` is a more Pythonic function name.

File: ciphers/porta_cipher.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+    if char in table[0]:
+        row = 0
+    elif char in table[1]:
+        row = 1
+
+    if row != -1:
+        return (row, table[row].index(char))
+    else:
+        return (None, None)
```
Comment (cclauss) [inline ciphers/porta_cipher.py]: Please convert these four lines into a one liner using a ternary if.
`return (None, None) if row == -1 else (row, table[row].index(char))`

File: ciphers/porta_cipher.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    'A'
+    """
+    row, col = getPosition(table, char.upper())
+
+    if row == 1:
+        return table[0][col]
+    elif row == 0:
+        return table[1][col]
+    else:
+        return char
```
Comment (cclauss) [inline ciphers/porta_cipher.py]: Please replace these four lines with:
```python
else:
    return table[1][col] if row == 0 else char
```

File: ciphers/porta_cipher.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    elif char in table[1]:
+        row = 1
+
+    if row != -1:
+        return (row, table[row].index(char))
+    else:
+        return (None, None)
+
+
+def getOpponent(table, char):
```
Comment (cclauss) [inline ciphers/porta_cipher.py]: get_opponent

File: ciphers/porta_cipher.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
+alphabet = {
```
Comment (cclauss) [inline ciphers/porta_cipher.py]: Please reformat this code with https://github.com/psf/black. Lines longer than 88 characters are going to get wrapped. 

Does this still work if these remain strings?
```python
     "A": ("ABCDEFGHIJKLM", "NOPQRSTUVWXYZ"),
```

File: ciphers/porta_cipher.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+def decrypt(key, words):
+    """
+    >>> decrypt('marvin', 'QRACRWU')
+    'JESSICA'
+    """
+    return encrypt(key, words)
+
+
+def getPosition(table, char):
```
Comment (cclauss) [inline ciphers/porta_cipher.py]: get_position

File: .github/workflows/directory_writer.yml [hunk 8]
```diff
@@ -0,0 +1,25 @@
+# The objective of this GitHub Action is to add a new DIRECTORY.md file to a pull request if needed.
```
Comment (cclauss) [inline .github/workflows/directory_writer.yml]: This file belongs in a different PR.

Comment (mrvnmchm) [reply .github/workflows/directory_writer.yml]: removed directory_writer.yml

File: ciphers/porta_cipher.py (hunk 6 above)
Comment (mrvnmchm) [reply ciphers/porta_cipher.py]: This works and formatting looks much better, added in latest changes

File: ciphers/porta_cipher.py (hunk 5 above)
Comment (mrvnmchm) [reply ciphers/porta_cipher.py]: completed in latest

File: ciphers/porta_cipher.py (hunk 4 above)
Comment (mrvnmchm) [reply ciphers/porta_cipher.py]: updated in latest

File: ciphers/porta_cipher.py (hunk 3 above)
Comment (mrvnmchm) [reply ciphers/porta_cipher.py]: updated in latest

File: ciphers/porta_cipher.py (hunk 2 above)
Comment (mrvnmchm) [reply ciphers/porta_cipher.py]: updated in latest

File: ciphers/porta_cipher.py [hunk 9]
```diff
@@ -1,107 +1,100 @@
...
+      """
+    return [alphabet[char] for char in key.upper()]
 
 
 def encrypt(key, words):
     """
-    >>> encrypt('marvin', 'jessica')
-    'QRACRWU'
-    """
+      >>> encrypt('marvin', 'jessica')
```
Comment (cclauss) [inline ciphers/porta_cipher.py]: This indenting is not needed or desirable.

File: ciphers/porta_cipher.py [hunk 10]
```diff
@@ -1,107 +1,100 @@
...
         count = (count + 1) % len(table)
     return cipher
 
 
 def decrypt(key, words):
     """
-    >>> decrypt('marvin', 'QRACRWU')
-    'JESSICA'
-    """
+      >>> decrypt('marvin', 'QRACRWU')
```
Comment (cclauss) [inline ciphers/porta_cipher.py]: Same.

File: ciphers/porta_cipher.py [hunk 11]
```diff
@@ -1,107 +1,100 @@
...
-    [[['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M'], ['U', 'V', 'W', 'X', 'Y', 'Z', 'N', 'O', 'P', 'Q', 'R', 'S', 'T']], [['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M'], ['N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z']], [['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M'], ['S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'N', 'O', 'P', 'Q', 'R']], [['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M'], ['Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'N', 'O', 'P']], [['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M'], ['W', 'X', 'Y', 'Z', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V']], [['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M'], ['U', 'V', 'W', 'X', 'Y', 'Z', 'N', 'O', 'P', 'Q', 'R', 'S', 'T']]]
-    """
-    table = []
-    for char in key.upper():
-        table.append(alphabet[char])
-    return table
+      >>> generate_table('marvin')
+      [('ABCDEFGHIJKLM', 'UVWXYZNOPQRST'), ('ABCDEFGHIJKLM', 'NOPQRSTUVWXYZ'), ('ABCDEFGHIJKLM', 'STUVWXYZNOPQR'), ('ABCDEFGHIJKLM', 'QRSTUVWXYZNOP'), ('ABCDEFGHIJKLM', 'WXYZNOPQRSTUV'), ('ABCDEFGHIJKLM', 'UVWXYZNOPQRST')]
```
Comment (cclauss) [inline ciphers/porta_cipher.py]: These long lines need to be wrapped.

File: ciphers/porta_cipher.py [hunk 12]
```diff
@@ -1,107 +1,100 @@
...
-    """
-    >>> table = [[['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M'], ['U', 'V', 'W', 'X', 'Y', 'Z', 'N', 'O', 'P', 'Q', 'R', 'S', 'T']], [['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M'], ['N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z']], [['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M'], ['S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'N', 'O', 'P', 'Q', 'R']], [['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M'], ['Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'N', 'O', 'P']], [['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M'], ['W', 'X', 'Y', 'Z', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V']], [['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M'], ['U', 'V', 'W', 'X', 'Y', 'Z', 'N', 'O', 'P', 'Q', 'R', 'S', 'T']]]
-    >>> getPosition(table, 'A')
-    (None, None)
+def get_position(table, char):
     """
+      >>> table = [('ABCDEFGHIJKLM', 'UVWXYZNOPQRST'), ('ABCDEFGHIJKLM', 'NOPQRSTUVWXYZ'), ('ABCDEFGHIJKLM', 'STUVWXYZNOPQR'), ('ABCDEFGHIJKLM', 'QRSTUVWXYZNOP'), ('ABCDEFGHIJKLM', 'WXYZNOPQRSTUV'), ('ABCDEFGHIJKLM', 'UVWXYZNOPQRST')]
```
Comment (cclauss) [inline ciphers/porta_cipher.py]: Same.

File: ciphers/porta_cipher.py [hunk 13]
```diff
@@ -1,107 +1,100 @@
...
-    >>> getOpponent(table, 'A')
-    'A'
+def get_opponent(table, char):
     """
-    row, col = getPosition(table, char.upper())
+      >>> table = [('ABCDEFGHIJKLM', 'UVWXYZNOPQRST'), ('ABCDEFGHIJKLM', 'NOPQRSTUVWXYZ'), ('ABCDEFGHIJKLM', 'STUVWXYZNOPQR'), ('ABCDEFGHIJKLM', 'QRSTUVWXYZNOP'), ('ABCDEFGHIJKLM', 'WXYZNOPQRSTUV'), ('ABCDEFGHIJKLM', 'UVWXYZNOPQRST')]
```
Comment (cclauss) [inline ciphers/porta_cipher.py]: Same.
